Mvc

good5874 opened the pull request:
enam не сделал с большой потому что потом капсом придётся в консоли отвечать)
думаю теперь стало ещё запутаннее)))

good5874 committed:
1

good5874 committed:
1.1

good5874 committed:
теперь пользователь работает с классом menu, там пакуется то что он ввёл и передаётся в класс managment, дальше происходит всё тоже самое

good5874 committed:
добавил проверку результатов

BitW13 commented on EpamTestConsole/Console/ConsoleCommandEnam.cs:
```diff
@@ -0,0 +1,4 @@
+﻿namespace EpamTestConsole
+{
+    enum Сommands { y, n, del }
```
{
y,
n,
del
}

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+            management.CreateNameTest(nameTest);
+        }  
+
+        private void OpenAndRunMethod(bool StartOrEdit)
```
startOrEdit

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+            management.CreateNameTest(nameTest);
+        }  
+
+        private void OpenAndRunMethod(bool StartOrEdit)
```
OpenAndRunMethod... Во-первых, если в названии метода сразу два глагола, это значит, что любое название не подходит, либо стоит разделить этот метод на два.

BitW13 replied:
Во-вторых не должно быть "..Method" в названии

good5874 committed:
добавил сохранение результатов теста, можно помотреть всю историю пройденых тестов или по названию отдельного теста.

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+            Console.WriteLine(ConsoleMenuConstant.EnterNamber);
+            switch (Console.ReadLine())
+            {
+                case "1":
```
Так не очень понятно, что за "1", "2" и т.д.
Подумай, как от этого избавиться - нужен самодокументированный код
Возможно здесь подойдет также enum

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+
+        public void StartConsole()
+        {            
+            Console.WriteLine("1. " + ConsoleMenuConstant.CreateTest);
```
А не получиться как-то совместить "1. " и ConsoleMenuConstant.CreateTest?
ConsoleMenuConstant.CreateTest где-то ещё используется и там не удобно будет, если добавиться "1. "? Если нет, тогда можно запихнуть в эту константу.
Проверь пожалуйста это для всех magic numbers

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+
+        private void CreateTest()
+        {
+            Console.WriteLine("Введите тему теста");
```
"Введите тему теста"
Если решился на то, чтобы запихивать всё в константы, в отдельный класс, то нужно продолжать так делать:)
Всегда помни о том, что в проекте всё должно придерживаться одного стиля.
Не обязательно, чтобы этот стиль был во всех проектах, главное, чтобы в пределах одного проекта - один стиль написания кода.

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+            Console.WriteLine("3. " + ConsoleMenuConstant.EditTest);
+
+            Console.WriteLine(ConsoleMenuConstant.EnterNamber);
+            switch (Console.ReadLine())
```
А если пользователь введет 4? или 5?) Или просто enter нажмет
Нужно это также учесть
Начни с того, чтобы в switch был default

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+                    Save(management);
+                    break;
+                case "2":
+                    OpenAndRunMethod(true);//start test
```
Я всё же думаю, что стоит разделить OpenAndRunMethod хотя бы на два метода.

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+                    break;
+                case "3":
+                    OpenAndRunMethod(false);//edit test
+                    Save(management);
```
А сохранение изменений (как я понял) должно быть внутри метода с логикой. UI класс не должен это решать

good5874 replied:
Я не понимаю зачем выносить этот метод в другой класс. Сохранение производит класс TestRepository. Save только выводит в консоль текст и принимает текст и вызывает метод который сохранит и передаёт параметры в него.

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+{
+    public class Menu
+    {
+        Management management = new Management();
```
Добавь пожалуйста везде модификаторы доступа

BitW13 commented on EpamTestConsole/Management.cs:
```diff
@@ -7,288 +7,99 @@ namespace EpamTestConsole
         {
-            Queue<TreeNode> q = new Queue<TreeNode>();
-            q.Enqueue(root);
+            RootNode.Section.NameSection = nameTest;                       
```
Лишние пробелы:)

BitW13 replied:
![image](https://user-images.githubusercontent.com/28789651/92995639-e561e700-f50d-11ea-9741-a1dcd82cf61a.png)
Я их вижу))

good5874 replied:
Хочу такое же супер зрение :)

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+            {
+                if (StartOrEdit)
+                {
+                    WalkTheTree(management.RootNode, WriteSectionToConsole);                    
```
Меня всё же смущает, что в какой-то момент мы перешли от тестов нодам)
Из кода не понятно, что такое Node и RootNode. Лучше наверное переименовать это на Test и RootTest соответственно

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+            {
+                case "1":
+                    CreateTest();
+                    WalkTheTree(management.RootNode, AddSection);
```
Я пока не до конца понимаю, что значит WalkTheTree... Лучше переименовать

BitW13 replied:
И как это связано с методами, которые пропихиваются в качестве параметров?

good5874 replied:
Какое подходящее название дать, есть идеи?)

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+                TreeNode temp = queue.Dequeue();
+                metod(temp);
+
+                if (temp.ChildNodes == null) continue;
```
Добавь фигурные скобки и внутри continue.
И такой вопрос... А можно как-нибудь обойтись здесь без continue?

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+            string _question = "";
+
+            string question;
+            string answer = "";            
```
Эти переменные можно объявлять в более узкой области жизни насколько я понял из кода.

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+
+        private void AddQuestion(TreeNode node)
+        {
+            string _question = "";
```
Зачем здесь в названии нижнее подчеркивание?

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+                    Console.WriteLine("Введите вопрос:");
+                    question = Console.ReadLine();
+                    Console.WriteLine("Введите ответ:");
+                    answer = Console.ReadLine();                    
```
Здесь есть лишние пробелы. Просмотри это везде.

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+                    Console.WriteLine("Введите ответ:");
+                    answer = Console.ReadLine();                    
+                                        
+                    answerOptions = AddAnswerOtions(question);
```
Может AddAnswerOptions

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+
+                    checkAnswer = false;
+                    if (answer != "")
+                        checkAnswer = true;
```
Добавь фигурные скобки

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+
+                    options = false;
+                    if (answerOptions != "")
+                        options = true;
```
И здесь

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+                        checkAnswer = true;
+
+                    options = false;
+                    if (answerOptions != "")
```
Можно проверить через string.IsNullOrEmpty

BitW13 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+                    if (answerOptions != "")
+                        options = true;
+                  
+                    _question += question + "\n" + checkAnswer + "\n" + options + "\n" + answer + "\n" + answerOptions;
```
Прочитай про переопределение метода ToString https://docs.microsoft.com/en-us/dotnet/api/system.object.tostring?view=netcore-3.1
Я думаю, идеально это зайдет, если ты его переопределишь в модели Question... да и наверное во всех моделях. Получится избавится от такого не очень красивого куска кода)

BitW13 commented on EpamTestConsole/Management.cs:
```diff
@@ -7,288 +7,99 @@ namespace EpamTestConsole
-            Queue<TreeNode> q = new Queue<TreeNode>();
-            q.Enqueue(root);
-            int i = 0;
+            var _question = question.Split('\n');
```
Меня смущает, что сперва ты лепишь в строку всё, а потом сплитишь)
Выше я написал, как лучше тебе это всё переделать под ToString. Поправь и возможно здесь вопрос со сплитом сразу решится

BitW13 commented on EpamTestConsole/Management.cs:
```diff
@@ -7,288 +7,99 @@ namespace EpamTestConsole
+            bool checkAnswer = false;
+            bool options = false;
+            
+            if(_question[1]=="True")
```
Тоже такое себе... ToString нужно юзать в случае чего и не привязываться к UI виду моделей

BitW13 commented on EpamTestConsole/Management.cs:
```diff
@@ -7,288 +7,99 @@ namespace EpamTestConsole
-            List<Question> Questions = new List<Question>();
+            var node = RootNode.Search(RootNode, nameSection);
+
+            var _question = question.Split('\n');            
```
Тоже самое....

BitW13 commented on EpamTestConsole/Management.cs:
```diff
@@ -7,288 +7,99 @@ namespace EpamTestConsole
 
-            Console.WriteLine($"Добавить к <{question.TextQuestion}> варианты ответов?");
+        public void CreateSection(string _sections)
+        {
```
Тоже самое

good5874 commented on EpamTestConsole/Console/Menu.cs:
```diff
@@ -0,0 +1,329 @@
+
+            Console.WriteLine(ConsoleMenuConstant.EnterNamber);
+            switch (Console.ReadLine())
+            {
```
Не понял что именно надо вынести

good5874 committed:
Поправивил Question, переопределил ToString, добавил в него проверку ответов
В menu теперь создаёт объекты question и section, а management добавляет их в нужный узел.

good5874 committed:
исправил опечатку

good5874 committed:
добавил проверку вопроса и вывод в консоль сразу после ответа на вопрос

good5874 committed:
Timer

good5874 committed:
Merge pull request #6 from good5874/CheckingAnswer

добавил проверку результатов
